### PR TITLE
feature: add collision handling for camera with terrain

### PR DESCRIPTION
**Pull request by nosy-b**
This PR adds a test for collision between the camera and the terrain (geometry layer) for all Views (Planar, Globe...) (issue #153).

### CODE REVIEW

File: src/Renderer/ThreeExtended/GlobeControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: this is actually `currentTile.OBB().pointsWorld()`

File: src/Renderer/ThreeExtended/GlobeControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: `if (currentTile.isElevationLayerLoaded())` should work.

File: src/Renderer/ThreeExtended/GlobeControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: Not a big deal, but I find it more readable to write:
```js
const image = currentTile.getLayerTextures(l_ELEVATION)[0].image;
const w = image.width;`
```
and later line 687
```js
const elevationUnderCamera = image.data[indiceElevation];
```

More readable, and *should* be more stable because internal storage could change, but I expect `getLayerTextures` to stay.

File: src/Renderer/ThreeExtended/GlobeControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: `elevationIndex`?

Comment (autra) [inline src/Renderer/ThreeExtended/GlobeControls.js]: In rigor, shouldn't it be 
```js
const indiceElevation = Math.floor(v * h) * w + Math.floor(u * w);
```
with `h = image.height`? (please check this though, I might have mixed up the axis).

In practice, I don't know if we'll have non-squared image, but I also find it easier to understand this way.

**Review by autra — DISMISSED** [review]
Looking mostly good :-) My comments are mainly cosmetics, except the last one (not sure if it has great consequences though).

**Comment by peppsac** [comment]
Adding features to `GlobeControls` (aka: globe only feature) is a bad idea, except for features that really requires a globe.

- we already have https://github.com/iTowns/itowns/blob/master/src/utils/DEMUtils.js to extract altitude information from textures. Please use/adapt it instead of adding new (globe specific) code.
- the collision with tile feature should probably live in `Camera` (or in another shared module) in order to be usable by everyone: PlanarView or View.

**Comment by nosy-b** [comment]
@peppsac Right, I'm working on it

**Comment by nosy-b** [comment]
@iTowns/reviewers Updated as generic method for all type of view with a geometry layer such as globe or planar

File: src/Core/View.js
Comment (peppsac) [inline src/Core/View.js]: I'm not sure all controls instances support external changes of camera's position.

File: src/Core/View.js
Comment (peppsac) [inline src/Core/View.js]: You can't really expect the first layer to always be the geometry one (even if it's true in GlobeView/PlanarView)

**Review by peppsac — CHANGES_REQUESTED** [review]
Now that this PR uses `DEMUtils` it's clearer that the code handling camera<->layer collision doesn't belong to `View`.

I can think of 2 alternatives:
- add a function to `Camera`: `adjustAltitudeToAvoidCollisionWithLayer` and use this function from controls. 
- or add collision handling code to controls classes (because they're in charge of camera's position)

In both cases, collision handling would be optional, and enabled by default in Globe/PlanarView.

WDYT?

**Comment by nosy-b** [comment]
@peppsac Agreed, the code location for collision handling is not ideal... but I liked the idea that any controls doesn't have to add a line to handle it. 
I change it now , commit very soon

**Comment by autra** [comment]
My 2 cents: let's enable it by default for our plane and globe examples while keeping in mind that some applications do *not* want collision detection (I'm working for an application using itowns, where some data are actually underground :-) ).

**Comment by nosy-b** [comment]
It's enabled by default for globeview and planarview but the default camera setting is set to not care about collision as we said.
@peppsac  WDYT? Ok for 2.1? Could be improved maybe later with smarter collision tracking with inertial speed variation for example.

File: src/Renderer/Camera.js
Comment (peppsac) [inline src/Renderer/Camera.js]: For now I'd prefer this option to exist only on the controls that support it.
So I'd add a `handleCollision` to `GlobeControls` and `PlanarControls` and some code in `PlanerView`/`GlobeView` constructor. For instance:
```
this.controls.handleCollision = options.handleCollision || false;
```

File: src/Renderer/Camera.js
Comment (peppsac) [inline src/Renderer/Camera.js]: You should pass the `layer` as a parameter to this function.
It should be more robust than assuming that it's the first `geometry` one.

Comment (peppsac) [reply src/Renderer/Camera.js]: So this also indicates that instead of storing a boolean `handleCollision` you probably want to store a list of layers that you want the camera to collide with.
So my earlier comment example code would become:
```
this.controls.collisionLayers = options.handleCollision ? [wgs84TileLayer] : [];
```

WDYT?

**Comment by nosy-b** [comment]
@peppsac it's not that simple as PlanarView doesn't reference its controller as opposed to GlobeView.
Nevertheless let's give it a try!

The collision options now sits in PlanarControls/GlobeControls. It's On by default but can be set in their respective options. While instantiating globview it is possibile to set it too.
This collision works with the (unique) geometry layer of the different views but it is now possible to pass any layer as a parameter. 
A very simple upgrade would make it work with an array of layers but to keep it clean I left it like that as we have a specific way to compute the elevation with elevation tiles and we might in the future check collision differently (against 3D tiles buildings for examples)

**Comment by peppsac** [comment]
You commited `Segmentation.js` which is unrelated to this PR :)

Also `this.controls.handleCollision = options.handleCollision || true;` will always result in `this.controls.handleCollision = true`...

**Comment by nosy-b** [comment]
Oups :)

**Comment by nosy-b** [comment]
@peppsac Are you ok for merge?

File: src/Renderer/Camera.js
Comment (peppsac) [inline src/Renderer/Camera.js]: `EPSG:4326` is wrong here (are you sure you tried in planar mode :) ?)

But the good news is: there's a ready to use function that does what you need: 
```js
const camLocation = view.camera.position(); 
```

File: src/Renderer/Camera.js
Comment (peppsac) [inline src/Renderer/Camera.js]: IMHO, either you:
- test for `undefined` and throw a useful error
- document that `elevationLayer` cannot be undefined

But testing and then ignoring `undefined` seems like a bad idea.

File: src/Renderer/Camera.js
Comment (peppsac) [inline src/Renderer/Camera.js:122]: cf: https://github.com/iTowns/itowns/pull/465#discussion_r138268173 
"I'm not sure all controls instances support external changes of camera's position."

Did you check?

Comment (peppsac) [reply src/Renderer/Camera.js:122]: You probably should use `view.notifyChange(...)` here since you modify the camera position.

**Review by peppsac — COMMENTED** [review]
Sorry but there a still a few things to address before merging.

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js]: No EPSG:4326 is geographic which is needed.

Comment (nosy-b) [reply src/Renderer/Camera.js]: (And yes of course I tested in planar and it works)

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js:122]: The three different controls we use support it.
But yes I thought in the future about having a function in camera that just compute a new position if collision so the controller itself can modify the camera position using this value.

File: src/Renderer/Camera.js
Comment (peppsac) [reply src/Renderer/Camera.js]: Good to know that it's working but... it doesn't make sense, right?
Why would you transform from a planar CRS to geographic to query elevation in a layer with a planar CRS?

File: src/Renderer/Camera.js
Comment (peppsac) [reply src/Renderer/Camera.js:122]: Ok, let's keep it this way if it's working fine with the available controls.

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js]: I understand your comment but here, it's not an error. If undefined it means there is no collision it's not a proper error

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js]: Yes it looks weird:) That's because we can easily manipulate the altitude in geographic coord using the coordinates class.
But yes I can simplify it to 
```javascript
const camLocation = view.camera.position().as('EPSG:4326');
```
and better, we could also modify Coordinates class so we handle altitude as a height if Coord is not geographic. This would avoid coordinate transform when in projected coordinate system

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js:122]: (We don't need view.notifyChange as this function is called on a control update which then notify the change)

File: src/Renderer/Camera.js
Comment (nosy-b) [reply src/Renderer/Camera.js]: I'll rebase and add a comment to the function so it's clearer

Comment (peppsac) [reply src/Renderer/Camera.js]: Ok, should be enough for now.

File: src/Renderer/Camera.js
Comment (peppsac) [reply src/Renderer/Camera.js]: IMHO it makes no sense to call this function without an elevationLayer.
So cluttering the code with unneeded `if X !== undefined` isn't bringing us any value.

So I'd prefer this function to have a clear contract ("I require an elevationLayer") and leave it to the user the task of not calling it with an invalid value.

File: src/Renderer/Camera.js
Comment (peppsac) [reply src/Renderer/Camera.js:122]: > We don't need view.notifyChange [...]

You need to call `view.notifyChange` because you're modifying the camera's position.
The fact that the controls update method may call `view.notifyChange` too is irrelevant.

Comment (nosy-b) [reply src/Renderer/Camera.js]: yes i did it in fact it's rebased

Comment (nosy-b) [reply src/Renderer/Camera.js:122]: yes i did it in fact it's rebased

**Review by peppsac — APPROVED** [review]
I still have a few concerns with this PR but I prefer to move forward and come back to this subject later.

It's a nice feature to have anyway so thanks for the work you put into it :)

**Pull request merged**
